ports/esp32/modwifi_csi.c

```diff
@@ -0,0 +1,705 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2025 Francesco Pace <francesco.pace@gmail.com>
+ * Copyright (c) 2025 MicroPython CSI Module Contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/mphal.h"
+#include "py/objarray.h"
+#include "py/runtime.h"
+
+#if MICROPY_PY_NETWORK_WLAN_CSI
+
+#include "esp_log.h"
+#include "esp_timer.h"
+#include "esp_wifi.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "modwifi_csi.h"
+#include <string.h>
+
+// Define STATIC if not already defined
+#ifndef STATIC
+#define STATIC static
+#endif
+
+static const char *TAG = "wifi_csi";
+
+// Global CSI state
+csi_state_t g_csi_state = {
+    .buffer =
+    {
+        .frames = NULL,
+        .head = 0,
+        .tail = 0,
+        .size = 0,
+        .dropped = 0,
+        .initialized = false,
+    },
+    .config =
+    {
+        .lltf_en = true,
+        .htltf_en = true,
+        .stbc_htltf2_en = true,
+        .ltf_merge_en = true,
+        .channel_filter_en = true,
+        .manu_scale = false,
+        .shift = 0,
+        .buffer_size = CSI_DEFAULT_BUFFER_SIZE,
+    },
+    .enabled = false,
+};
+
+// ============================================================================
+// Circular Buffer Implementation (Lock-Free)
+// ============================================================================
+
+static bool csi_buffer_init(csi_buffer_t *buf, uint32_t size) {
+    if (buf->initialized) {
+        // Already initialized, free old buffer
+        if (buf->frames) {
+            free(buf->frames);
+        }
+    }
+
+    buf->frames = (csi_frame_t *)malloc(sizeof(csi_frame_t) * size);
```

Comment:
If this buffer doesn't have to be part of the ESP-IDF heap or you have some special requirements on how this buffer's lifetime is managed, then I guess you should probably use MicroPython's memory management functions instead.

In this case you could probably use `m_new` and then wrap the call to `csi_buffer_init` in an NLR context (the try-catch mechanism used by native code), or `m_new_maybe`.  If you use MicroPython's memory management functions, all instances of `free` instances should turn into `m_del`.